### ISSUE TITLE
Plugin is broken on install and requires to restart Godot due to resources failing to load

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#23937]
Godot 3.1 alpha2

I copied a working addon into a Godot 3.1 project, but it was completely broken and produced a lot of errors. I restarted the editor, the errors went away and the plugin worked.
The same plugin in Godot 3.0.6 worked fine without errors, even on first install.
I tried to reduce the size of the project to narrow it down, but it also changed the results and became not reproducible. It always happens if I test with the whole plugin. It's a small plugin though, with two scenes.

Reproduction steps:

1) Download this project: [AddonBrokenWhenFirstAdded.zip](https://github.com/godotengine/godot/files/2611610/AddonBrokenWhenFirstAdded.zip)

2) Open it in Godot. It's an empty project as if you had created it from scratch, and should load fine.

3) Go to the system file explorer and unzip the `addons` archive to install the addon (works too if you created an empty project from scratch).

4) Focus back the editor, see it detects and imports the addon's resources. However, it prints a lot of errors which didn't happen in Godot 3.0.6. Most notably this error, which causes a lot of others due to the script not being compiled:
```
SCRIPT ERROR: GDScript::reload: Parse Error: Can't preload resource at path: res://addons/zylann.channel_packer/icons/empty.png
          At: res://addons/zylann.channel_packer/input_texture.gd:14
```
This resource definitely exists.

5) Go to project settings and activate the plugin: it makes a new menu appear on top-right, use it, notice the dialog it opens is completely broken:
![image](https://user-images.githubusercontent.com/1311555/48959221-f10ca400-ef5b-11e8-9d20-096f17f0298c.png)
Note: the layout problem is caused by https://github.com/godotengine/godot/issues/17626

6) Close the project, reopen it: notice everything went back to normal.
![image](https://user-images.githubusercontent.com/1311555/48959302-5eb8d000-ef5c-11e8-9036-a04dfc2fe6a4.png)

**Comment by Zylann** [comment]
This problem also happens if you want to cleanup the `import` folder. This folder never collects garbage so it keeps growing to insane sizes, the only solution being to delete it.

**Comment by avencherus** [comment]
For my projects I .gitignore the imports folder, so I don't have that issue.  Since the editor generates what is needed, I'm free to erase the folder and allow it to regenerate without interfering with the version.

I've been at least assuming it was never intended to be versioned.  Similar to the editor thumbnails in the user data folder.

**Comment by Zylann** [comment]
> For my projects I .gitignore the imports folder, so I don't have that issue. 

So do I. That is precisely what makes this situation occur, so you may have been lucky.

**Comment by avencherus** [comment]
Ok with those revisions I think I understand better. The imports in the .import folder also have to be present at startup for a plugin to load certain resources.

**Comment by clemens-tolboom** [comment]
> No loader found for resource: res://addons/gdUnit4/src/ui/assets/PlayDebug.svg

Just did #44940 (removing imports) but on v4.0.3.stable.official [5222a99f5]

**Comment by KoBeWi** [comment]
The issue should've been fixed by #70668
Does anyone have up-to-date MRP?

**Comment by MikeSchulze** [comment]
Hi, to bring this topic forward here are my class loading issues from my plugin.
Maybe it is more related to https://github.com/Zylann/godot_heightmap_plugin/issues/26

I have minimized the plugin to show the problems when the plugin is first restarted.
Using `v4.2.1.stable.official [b09f793f5]`
[plugin_class_load_bug.zip](https://github.com/godotengine/godot/files/13927689/plugin_class_load_bug.zip)

As far as I understand, the main problem is the use of classes by name, the classes are not loaded at plugin start time and so are not found.

Open the attached Project and check the errors.
```
Godot Engine v4.2.1.stable.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
  modules/gltf/register_types.cpp:63 - Blend file import is enabled in the project settings, but no Blender path is configured in the editor settings. Blend files will not be imported.
--- Debug adapter server started ---
--- GDScript language server started on port 6005 ---
  res://addons/gdUnit4/plugin.gd:17 - Parse Error: Identifier "GdUnitSettings" not declared in the current scope.
  modules/gdscript/gdscript.cpp:2788 - Failed to load script "res://addons/gdUnit4/plugin.gd" with error "Parse error". (User)
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_green.svg-9ce01031b489dea619e91196cb66dce9.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red1.svg-8ddad0e8a9c8884f19621c8f733ce341.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red2.svg-3ab7610a37b37b2e46ca7faaba5a2c40.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red3.svg-2f1ae0a474446c730d13e401878da7f2.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red4.svg-b6aa919f74c7c5f9e6a5a87210e2194e.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red5.svg-94eac4c39a2d4020f502eab72982b3f2.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red6.svg-fb927f6dc4442199133d2e25e9d9d21a.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red7.svg-4b4ab8aec1bc8343d5df6b64a24f7c22.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseError1.svg-43d13ea25f9f8c66fecf5a0ab4a752ad.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseError2.svg-27dc52b88f226d741b1f9c1294295841.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed1.svg-32464e8f6fa7f74ad74a7534dfaba019.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed2.svg-aa38e10f09edf43b31b1c0a4caf549c5.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed.svg-151182f42f6b32bd8c6dc168e9469b54.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseSuccess1.svg-3eadebb620a3275b67f53d505bc0f96b.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseSuccess2.svg-7767c6ebe7bd14d031b8c87b24e08595.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress1.svg-baca226eb5c6ca50a0b5f3af77fe615c.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress2.svg-6a0cbcb42a8df535c533cf79599952d6.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress3.svg-0b465f11e95f98f7b157a0bf0ded40c1.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress4.svg-09def7d3fee66ec2764c9bbd72c3a961.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress5.svg-9874b4bd1c734fd859a28d95960c17c5.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress6.svg-0a3b2b954e3a285cee1f29222aac701b.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress7.svg-e824844cb9cfdf076f9196cf47098a7d.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress8.svg-b37d84176a257f378f0f5dff81bfc322.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/clock.svg-b16f5d68e1dedc017f1ce1df1e590248.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/errors.svg-53aa38a5c5d3309095cd34f36952f16b.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/failures.svg-8659a946adf9b0616e867a8bf0855d3d.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/PlayDebug.svg-d3618ec14e2e4cb6b467c3249916f8dd.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/PlayOverall.svg-d07157735d6bab5d74465733e8213328.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCase.svg-f6ee172ad0e725d3612bec1b6f3c8078.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseError.svg-373307086979f3f0e012eb3660cc91ec.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed.svg-df47525fd14d5e4149690cacd8eb08db.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseSuccess.svg-aaf852c6aeda68c93a410c7480502895.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestSuite.svg-f3ba31540dedae19e6c1b7b050a1b5d7.editor.meta'
GdUnit4 plugin loaded
```

The attempt to preload the class ended in a worst-case scenario and more and more errors occurred.
* Edit the plugin.gd and uncomment this lines.
```
# included this lines to try a workaround to preload missing class at plugin first start time
#@warning_ignore("shadowed_global_identifier")
#const GdUnitSettings := preload("res://addons/gdUnit4/src/core/GdUnitSettings.gd")
```
* Close the editor and delete the .godot folder
* Restart and see the errors
```
Godot Engine v4.2.1.stable.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
  modules/gltf/register_types.cpp:63 - Blend file import is enabled in the project settings, but no Blender path is configured in the editor settings. Blend files will not be imported.
--- Debug adapter server started ---
--- GDScript language server started on port 6005 ---
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:230 - Parse Error: Could not find type "GdUnitProperty" in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:253 - Parse Error: Could not find type "GdUnitProperty" in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:265 - Parse Error: Could not find type "GdUnitProperty" in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:271 - Parse Error: Could not find type "GdUnitProperty" in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:304 - Parse Error: Could not find type "GdUnitProperty" in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:120 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:120 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:121 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:121 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:122 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:122 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:123 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:123 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:125 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:125 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:126 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:126 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:127 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:127 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:129 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:129 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:130 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:130 - Parse Error: Identifier "GdUnitShortcut" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:231 - Parse Error: Could not find type "GdUnitProperty" in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:239 - Parse Error: Identifier "GdUnitProperty" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:260 - Parse Error: Identifier "GdUnitSignals" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:267 - Parse Error: Identifier "GdUnitSignals" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:312 - Parse Error: Identifier "GdUnitProperty" not declared in the current scope.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:317 - Parse Error: Cannot infer the type of "property" variable because the value doesn't have a set type.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:321 - Parse Error: Invalid get index "value" on base Nil
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:321 - Parse Error: Function "value()" not found in base null.
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:321 - Parse Error: Invalid get index "value" on base Nil
  res://addons/gdUnit4/src/core/GdUnitSettings.gd:321 - Parse Error: Function "value()" not found in base null.
  res://addons/gdUnit4/plugin.gd:-1 - Compile Error: 
  modules/gdscript/gdscript.cpp:2788 - Failed to load script "res://addons/gdUnit4/plugin.gd" with error "Compilation failed". (User)
  res://addons/gdUnit4/plugin.gd:17 - Invalid call. Nonexistent function 'setup' in base 'GDScript'.
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_green.svg-9ce01031b489dea619e91196cb66dce9.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red1.svg-8ddad0e8a9c8884f19621c8f733ce341.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red2.svg-3ab7610a37b37b2e46ca7faaba5a2c40.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red3.svg-2f1ae0a474446c730d13e401878da7f2.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red4.svg-b6aa919f74c7c5f9e6a5a87210e2194e.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red5.svg-94eac4c39a2d4020f502eab72982b3f2.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red6.svg-fb927f6dc4442199133d2e25e9d9d21a.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/orphan_red7.svg-4b4ab8aec1bc8343d5df6b64a24f7c22.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseError1.svg-43d13ea25f9f8c66fecf5a0ab4a752ad.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseError2.svg-27dc52b88f226d741b1f9c1294295841.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed1.svg-32464e8f6fa7f74ad74a7534dfaba019.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed2.svg-aa38e10f09edf43b31b1c0a4caf549c5.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed.svg-151182f42f6b32bd8c6dc168e9469b54.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseSuccess1.svg-3eadebb620a3275b67f53d505bc0f96b.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseSuccess2.svg-7767c6ebe7bd14d031b8c87b24e08595.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress1.svg-baca226eb5c6ca50a0b5f3af77fe615c.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress2.svg-6a0cbcb42a8df535c533cf79599952d6.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress3.svg-0b465f11e95f98f7b157a0bf0ded40c1.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress4.svg-09def7d3fee66ec2764c9bbd72c3a961.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress5.svg-9874b4bd1c734fd859a28d95960c17c5.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress6.svg-0a3b2b954e3a285cee1f29222aac701b.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress7.svg-e824844cb9cfdf076f9196cf47098a7d.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/Progress8.svg-b37d84176a257f378f0f5dff81bfc322.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/clock.svg-b16f5d68e1dedc017f1ce1df1e590248.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/errors.svg-53aa38a5c5d3309095cd34f36952f16b.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/failures.svg-8659a946adf9b0616e867a8bf0855d3d.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/PlayDebug.svg-d3618ec14e2e4cb6b467c3249916f8dd.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/PlayOverall.svg-d07157735d6bab5d74465733e8213328.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCase.svg-f6ee172ad0e725d3612bec1b6f3c8078.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseError.svg-373307086979f3f0e012eb3660cc91ec.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseFailed.svg-df47525fd14d5e4149690cacd8eb08db.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestCaseSuccess.svg-aaf852c6aeda68c93a410c7480502895.editor.meta'
  Missing required editor-specific import metadata for a texture (please reimport it using the 'Import' tab): 'res://.godot/imported/TestSuite.svg-f3ba31540dedae19e6c1b7b050a1b5d7.editor.meta'
```

Remark:
Deleting the `.godot` folder is like a repository checkout.
To go more into the plugin problems.
Since there is no update plugin mechanism, users are forced to update the plugins manually. This is done by manually deleting and replacing the addon folder with a new version. Often it is also necessary to delete the .godot folder due to existing caching bugs.

**Comment by KoBeWi** [comment]
Should be fixed by #92667
